enhancements/security/manual-cloud-credentials.md

```diff
@@ -0,0 +1,170 @@
+and make progress.
+
+  * Providing a mechanism to disable the cloud credential operator, this mechanism already exists.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints [optional]
+
+This proposal requires the CCO to begin reporting Upgradable=False in a number of scenarios, unless an admin informs us they have taken action to reconcile new permissions and it is safe to upgrade.
+
+CLI tooling will be implemented to help users who wish to disable the CCO to perform their initial installations, update credentials prior to upgrade for release images available for update, and reconcile credentials against current CredentialsRequests in the cluster on an on-going basis. Users must take this manual step to clear the Upgradable=False condition if credentials have changed in any available update.
+
+Cloud credential operator reconcile code will be refactored to be a little more friendly to use as a vendored library, for use in the CLI.
+
+Note that when we say CCO is disabled, it is still running, just not minting credentials, and does not have access to the powerful "root" credentials for the cloud provider account.
+
+### Risks and Mitigations
+
+
+## Design Details
+
+Core of this proposal will be CLI tooling which vendors code from the credentials operator and allows for single use execution of the reconcile loop. Provising CLI tooling allows users to bypass the security concerns of storing powerful cloud credentials in the cluster and instead keep them entirely locally.
+
+### Upgradable False Condition
+
+A field will be added to the cloud credential operator ConfigMap (possibly becoming a separate API type in another enhancement) storing a list of Updates (from ClusterVersion.AvailableUpdates) that have been reconciled, confirming that we have minted credentials and it is safe to upgrade to this release.
+
+```yaml
+  reconciledAvailableUpdates:
+  - version: 4.4.1
+    image: quay.io/openshift-release-dev/ocp-release:4.4.1-x86_64
+  - version: 4.4.2
+    image: quay.io/openshift-release-dev/ocp-release:4.4.2-x86_64
+```
+
+Entries will be added to this list by the CLI tooling below, however an administrator could do this by hand if desired.
+
+If the CCO is enabled per normal best practice install, this field will be ignored, and should be absent.
+
+However if CCO is disabled, it will compare the ClusterVersion.Spec.AvailableUpdates to this list of ReconciledAvailableUpdates. If *any* available update is not present in the reconciled list, the CCO will report Upgradable=False.
+
+### Initial Installation
+
+```bash
+$ oc adm credentials reconcile --manifests=/path/to/installer/manifests
+```
+
+This command would parse the CredentialsRequests in the manifests directory, mint credentials in the cloud using the local users environment variables or credentials file, store the status on the CredentialsRequsts (which are yet to be created), write Secret manifests containing the cloud credentials, and disable the CCO.
```

Comment:
> This command would parse the CredentialsRequests in the manifests directory

hmm.. who put the CredentialsRequests in the manifests directory in the first place, aren't the CredentialRequests in the relase-image. Also how will it decide which platform it needs to reconcile for incase it looks up the release image?

Reply:
> ... aren't the CredentialRequests in the relase-image...

The installer already vendors cred-operator tooling that [effectively inlines these requests][1].  Inlining the `CredentialsRequests` themselves doesn't seem like that much of a regression.  Personally I'd rather handle these on the bootstrap machine to decouple the installer code from the release image, but I understand that "I guess we can try pass-through... Launch the bootstrap machine... Ah, nope, guess we didn't have enough creds!" is not going to be the best UX.  It's possible that you could handle that by recommending folks using passthrough creds or a disabled cred-operator wash things through the `oc adm credentials ...` helper first.  Or maybe have this helper look up the target pullspec, pull it down, and create the `CredentialsRequests` itself.  Or something...

Also nit: `CredentialsRequsts` -> `CredentialsRequests`.

[1]: https://github.com/openshift/installer/blob/f2e197038c59417020ab1781e8a0d4a9e3d7cc28/vendor/github.com/openshift/cloud-credential-operator/pkg/aws/utils.go#L40-L92

Reply:
I see I may be confused about something here, I thought release image contents were in the manifests directory, I am guessing now that must not be the case. 

I will revise to avoid any edits of the CredentialsRequests to store metadata. It's possible we could store it on the secrets themselves, it's also possible we don't need to store anything, just the existence of the secret at the target location may be sufficient. Need to think on this a little more.

Reply:
Yes, just creating the secret with the creds that CCO would have done (if it were enabled) is enough to get image-registry, machine-api, ingress-operator up and running.

Reply:
The problem here is storing the state if we auto-reconciled using CLI tooling, if we can't store that state on a cred request. (state being the aws username, azure service principal, app ID, and gcp service account ID) 

I'm toying with the idea of serializing a full CredRequestStatus to json and storing as an annotation on the secret we drop into the manifests. Cred minter would then detect that, move the state on the actual cred request, and from that point onward we're ok as we have a live CredentialsRequest we can interact with.

Reply:
at least with AWS, you can go from ACCESS_KEY/SECRET to IAM username with an iam:GetUser call.

Reply:
and it seems like with GCP and Azure the user name/service principal is part of the secret payload